lib/src/rules/voidness_propagation.dart

```diff
@@ -0,0 +1,165 @@
+
+  @override
+  void visitExpressionFunctionBody(ExpressionFunctionBody node) {
+    DartType returnType;
+    if (node?.parent?.parent is FunctionDeclaration) {
+      returnType = (node.parent.parent as FunctionDeclaration).returnType?.type;
+    } else if (node?.parent is MethodDeclaration) {
+      returnType = (node.parent as MethodDeclaration).returnType?.type;
+    }
+    _check(returnType, node.expression.staticType, node.expression);
+  }
+
+  @override
+  void visitInstanceCreationExpression(InstanceCreationExpression node) {
+    final args = node.argumentList.arguments;
+    final parameters = node.staticElement?.parameters;
+    if (parameters != null) {
+      _checkArgs(args, parameters);
+    }
+  }
+
+  @override
+  void visitMethodDeclaration(MethodDeclaration node) {
+    final inheritedMethod = DartTypeUtilities.lookUpInheritedMethod(node);
+    if (inheritedMethod != null) {
+      _check(
+          inheritedMethod.returnType, node.returnType?.type, node.returnType);
+    } else {
+      for (final e in DartTypeUtilities.overridenMethods(node)) {
+        _check(e.returnType, node.returnType?.type, node.returnType);
+      }
+    }
+  }
+
+  @override
+  void visitMethodInvocation(MethodInvocation node) {
+    final type = node.staticInvokeType;
+    if (type is FunctionType) {
+      final args = node.argumentList.arguments;
+      final parameters = type.parameters;
+      _checkArgs(args, parameters);
+    }
+  }
+
+  @override
+  void visitReturnStatement(ReturnStatement node) {
+    final parent = node.getAncestor((e) =>
+        e is FunctionExpression ||
+        e is MethodDeclaration ||
+        e is FunctionDeclaration);
+    if (parent is FunctionExpression) {
+      final type = parent.staticType;
+      if (type is FunctionType) {
+        _check(type.returnType, node.expression?.staticType, node.expression);
+      }
+    } else if (parent is MethodDeclaration) {
+      _check(parent.declaredElement.returnType, node.expression?.staticType,
+          node.expression);
+    } else if (parent is FunctionDeclaration) {
+      _check(parent.declaredElement.returnType, node.expression?.staticType,
+          node.expression);
+    }
+  }
+
+  @override
+  void visitVariableDeclaration(VariableDeclaration node) {
+    if (node.initializer != null) {
+      _check((node.parent as VariableDeclarationList).type?.type,
+          node.initializer.staticType, node.initializer);
+    }
+  }
+
+  void _check(DartType expectedType, DartType type, AstNode node) {
+    if (expectedType == null || type == null) {
+      return;
+    } else if (expectedType is FunctionType && type is FunctionType) {
+      _check(expectedType.returnType, type.returnType, node);
+    } else if (type is InterfaceType &&
+        expectedType is InterfaceType &&
+        type != expectedType &&
+        type.element == expectedType.element) {
+      for (var i = 0; i < type.typeArguments.length; i++) {
+        if (type.typeArguments[i].isVoid &&
```

Comment:
I would call `_check` on `typeArguments[i]`.

I think the current code misses this:

```
void Function() x;
int Function() y = x;
```